set result in ComposedLastHttpContent instead of being null

KowalczykBartek opened the pull request:
Motivation:
I want to fix bug in vert.x project (https://github.com/eclipse-vertx/vert.x/issues/2562) caused by ComposedLastHttpContent result being null. I don't know if it is intentional that this last decoded chuck in the issue returns null, but if not - I am providing fix for that.

Modification:
1. Added new constructor in ComposedLastHttpContent allowing to pass DecoderResult
2. set DecoderResult.SUCCESS for created ComposedLastHttpContent in HttpContentEncoder
3. set DecoderResult.SUCCESS for created ComposedLastHttpContent in HttpContentDecoder

Result:
Fixes https://github.com/eclipse-vertx/vert.x/issues/2562 

cheers!

normanmaurer commented:
@netty-bot test this please

KowalczykBartek commented:
I added checks to existing tests (I expect that I will receive last LastHttpContent which contains not null result, when there is trailing headers present). Problem is that, there is no corresponding test for HttpContentDecompressorTest (by design ?), should I add such new test class for it also ?

normanmaurer commented:
@netty-bot test this please

KowalczykBartek commented:
@normanmaurer so, should I add test for decompression/decoding ? because I see that you approved this already.

normanmaurer commented:
@KowalczykBartek if you can sure... that said if you are too busy its fine as well and we can do a followup.

KowalczykBartek commented:
No, I will do it today, please wait with merging.

normanmaurer commented:
@KowalczykBartek let me know once done as I would like to include this in the next release if possible

KowalczykBartek commented:
@normanmaurer I created test (that was challenging :D ). The only problem with that test is ... It do not pass, because my trailer is missing - after some debugging I noticed that it(trailer) is correctly parsed but at the end is ignored :( 
HttpObjectDecoder in line 675 is decoding trailer, then it is trying to parse headers (?), if headerParsers return null, it ignores my trailer. 
```java
line = headerParser.parse(buffer);
if (line == null) {
    return null;
}
```

I modified slightly this code, 
```java
line = headerParser.parse(buffer);
if (line == null) {
    if(!trailer.trailingHeaders().isEmpty())
    {
        this.trailer = null;
        return trailer;
    }
    return null;
}
```
and my test pass - but for sure this change breaks other stuff :D I will try to understand HttpObjectDecoder, but maybe you see mistake that I done in my test?

KowalczykBartek commented:
yeah ! I found what was wrong - and pushed the change ! 

Question:
Can I say that it is a bug when HttpObjectDecoder cannot parse that:
```
assertTrue(channel.writeInbound(Unpooled.copiedBuffer("My-Trailer: 42", CharsetUtil.US_ASCII)));
assertTrue(channel.writeInbound(Unpooled.copiedBuffer("\r\n", CharsetUtil.US_ASCII)));
assertTrue(channel.writeInbound(Unpooled.copiedBuffer("\r\n\r\n", CharsetUtil.US_ASCII)));
````
but can parse that:
```
assertTrue(channel.writeInbound(Unpooled.copiedBuffer("My-Trailer: 42\r\n\r\n\r\n", CharsetUtil.US_ASCII)));
````

normanmaurer commented:
Yes... please open an issue

> Am 19.01.2019 um 19:42 schrieb Bartek Kowalczyk <notifications@github.com>:
> 
> yeah ! I found what was wrong - and pushed the change !
> 
> Question:
> Can I say that it is a bug when HttpObjectDecoder cannot parse that:
> 
> assertTrue(channel.writeInbound(Unpooled.copiedBuffer("My-Trailer: 42", CharsetUtil.US_ASCII)));
> assertTrue(channel.writeInbound(Unpooled.copiedBuffer("\r\n", CharsetUtil.US_ASCII)));
> assertTrue(channel.writeInbound(Unpooled.copiedBuffer("\r\n\r\n", CharsetUtil.US_ASCII)));
> but can parse that:
> 
> assertTrue(channel.writeInbound(Unpooled.copiedBuffer("My-Trailer: 42\r\n\r\n\r\n", CharsetUtil.US_ASCII)));
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

KowalczykBartek commented:
I created an issue : https://github.com/netty/netty/issues/8736

normanmaurer commented:
@KowalczykBartek please fix the checkstyle errors:

13:35:18 /code/codec-http/src/test/java/io/netty/handler/codec/http/HttpContentDecoderTest.java:104:52: Unnecessary parentheses around expression.
13:35:18 /code/codec-http/src/test/java/io/netty/handler/codec/http/HttpContentDecoderTest.java:106: Line is longer than 120 characters (found 139).
13:35:18 Audit done.
13:35:18 

Also please ensure you release all messages that implement `ReferenceCounted`

KowalczykBartek commented:
Fixed checkstyle errors (now no errors after mvn clean package) - btw sorry I forget about code-style :D 

> Also please ensure you release all messages that implement ReferenceCounted

Fixed, from what I understand only DefaultHttpContent disposable content.

normanmaurer commented:
@netty-bot Test this please

normanmaurer commented:
@KowalczykBartek thanks a lot!
